ci: Update Windows task image up to `visualstudio2022`

hebasto opened the pull request:
Besides upgrading Visual Studio, which seems [inevitable](https://github.com/bitcoin/bitcoin/pull/24531#discussion_r887854193), this PR also:
- bumps vcpkg to the latest version (previous one was in bitcoin/bitcoin#24847)
- fixes cache size limit for `ccache`

hebasto committed:
ci: Update Windows task image up to visualstudio2022

MarcoFalke commented on .cirrus.yml:
```diff
@@ -99,10 +99,10 @@ task:
     cpu: 4
     memory: 8G
-    image: cirrusci/windowsservercore:visualstudio2019
+    image: cirrusci/windowsservercore:visualstudio2022
```
I can't find this in the docs: https://cirrus-ci.org/guide/windows/#os-versions

MarcoFalke commented:
Looks like it is now?

MarcoFalke commented:
Please revert fa72e0ba15c6382e9068be221ab4872bef000cbc as well

hebasto commented:
> Looks like it is now?

Still getting "Agent is not responding!" error (

hebasto commented:
@sipsorcery 

Could you look at linker error here? (fwiw, I have wiped all caches)

MarcoFalke commented on .cirrus.yml:
```diff
@@ -99,10 +99,10 @@ task:
   windows_container:
```
could adjust the name to mention 22?

hebasto replied:
Done.

hebasto committed:
ci: Limit ccache cache size properly on "Win64 native" task

hebasto committed:
ci: Bump vcpkg to the latest version

Dependency changes:
 - boost-* 1.78.0#0 -> 1.79.0#0

hebasto committed:
ci: Increase CPU number for "Win64 native" task

Currently, the time it takes to get the "Win64 native" task done with
all of the caches been invalidated is very close to the 2 hours limit.

This task is the only one which runs on Windows Community Cluster,
therefore this change should not affect other CI tasks.

hebasto committed:
doc: Make Windows build docs match the CI task

Added Visual Studio 2022.

hebasto commented:
@MarcoFalke 
> Please revert [fa72e0b](https://github.com/bitcoin/bitcoin/commit/fa72e0ba15c6382e9068be221ab4872bef000cbc) as well

See #25472.

MarcoFalke commented on .cirrus.yml:
```diff
@@ -94,17 +94,17 @@ task:
   << : *FILTER_TEMPLATE
   windows_container:
-    cpu: 4
+    cpu: 6
```
any reason to change this? With 6 cpu it finished 1 minute earlier compared to master ( https://cirrus-ci.com/task/5018066990923776 ), so it seems like a random fluke.

Also, if someone pushes several pull requests at the same time, the scheduler will hit the total cpu limit earlier.

MarcoFalke replied:
Also, looking at the graph, the CPU usage density is less, indicating that the 5th and 6th cpu may be idle most of the time

hebasto replied:
> any reason to change this?

On master, the time required to get the task done with no caches is very close to 2 hours limit.

> With 6 cpu it finished 1 minute earlier compared to master ( https://cirrus-ci.com/task/5018066990923776 ), so it seems like a random fluke.

Such comparison is not correct, because `vcpkg_binary` cache was invalidated for this PR. Functional tests are more then 10 minutes faster now.

> Also, looking at the graph, the CPU usage density is less, indicating that the 5th and 6th cpu may be idle most of the time

Will 5 cpus be OK?

MarcoFalke replied:
> On master, the time required to get the task done with no caches is very close to 2 hours limit.

Ok, no objection then.

Is there a reason why vs22 needs more resources?

hebasto replied:
> Is there a reason why vs22 needs more resources?

On master, with [cleared caches](https://cirrus-ci.com/task/5306029448626176):
- time: 01:51:03
- peak memory usage: 5.10 GB

![Screenshot from 2022-06-26 12-27-08](https://user-images.githubusercontent.com/32963518/175809957-f7966d36-1f48-4c4d-a919-f4b32679038b.png)

With 156bc89788034c509051e0635ed40694b8d717c6 commit, with [cleared caches](https://cirrus-ci.com/task/6092886604251136):
- _\<TBD>_

hebasto replied:
Actually, there is no significant difference in resources at all.

MarcoFalke replied:
I think it might be best to revert this. Looks like now this won't be scheduled until 4 hours later. If the CPU was reduced, so that it was scheduled earlier, it would take 15 minutes longer to run, but still finish earlier.

hebasto replied:
> I think it might be best to revert this. Looks like now this won't be scheduled until 4 hours later. If the CPU was reduced, so that it was scheduled earlier, it would take 15 minutes longer to run, but still finish earlier.

Ok, But a similar situation with ARM task been observed right now.

hebasto replied:
> I think it might be best to revert this.

#25483

MarcoFalke commented:
too bad that this doesn't fix https://github.com/bitcoin/bitcoin/pull/25472. Though, I guess it can't hurt to bump now, given that we'll have to do it anyway at some point. However, I think our windows docs should match the windows CI.

sipsorcery commented:
> @sipsorcery
> 
> ~Could you look at linker error here? (fwiw, I have wiped all caches)~
> 
> nm, forgot to update `PlatformToolset` :)

I find myself doing that everytime I do a build. Maybe this PR is a good time to udate the default?

sipsorcery commented:
ACK 3dabd2891825a0b9a40276dffc5953f91a013c8d.

hebasto committed:
build: Bump default `PlatformToolset` for Visual Studio 2022

MarcoFalke commented on .cirrus.yml:
```diff
@@ -179,7 +179,7 @@ task:
     - '%x64_NATIVE_TOOLS%'
     - cd %CIRRUS_WORKING_DIR%
-    - ccache --zero-stats
+    - ccache --zero-stats --max-size=%CCACHE_SIZE%
```
is `CCACHE_SIZE` set at all?

hebasto replied:
https://github.com/bitcoin/bitcoin/blob/0dd34773334c7f4db7b05df30ee61b011795b46d/.cirrus.yml#L6

hebasto replied:
On [master](https://api.cirrus-ci.com/v1/task/5018066990923776/logs/build.log):
```
Summary:
  Hits:             239 /  239 (100.0 %)
    Direct:         239 /  239 (100.0 %)
    Preprocessed:     0 /    0
  Misses:             0
    Direct:           0
    Preprocessed:     0
  Uncacheable:       10
Primary storage:
  Hits:             478 /  478 (100.0 %)
  Misses:             0
  Cache size (GB): 1.35 / 5.00 (27.06 %)

Use the -v/--verbose option for more details.
```

With [this PR](https://api.cirrus-ci.com/v1/task/5900879822848000/logs/build.log):
```
Summary:
  Hits:             239 /  239 (100.0 %)
    Direct:         239 /  239 (100.0 %)
    Preprocessed:     0 /    0
  Misses:             0
    Direct:           0
    Preprocessed:     0
  Uncacheable:       10
Primary storage:
  Hits:             478 /  478 (100.0 %)
  Misses:             0
  Cache size (GB): 0.07 / 0.20 (32.93 %)

Use the -v/--verbose option for more details.
```

Compare the "Cache size" string.

hebasto commented:
Updated 3dabd2891825a0b9a40276dffc5953f91a013c8d -> 05b2d9fe076233961ebf24e74d47c57cc0618bbb ([pr25460.01](https://github.com/hebasto/bitcoin/commits/pr25460.01) -> [pr25460.02](https://github.com/hebasto/bitcoin/commits/pr25460.02), [diff](https://github.com/hebasto/bitcoin/compare/pr25460.01..pr25460.02)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/25460#issuecomment-1166445088):
> However, I think our windows docs should match the windows CI.

- addressed @sipsorcery's [comment](https://github.com/bitcoin/bitcoin/pull/25460#issuecomment-1166455372):
>  >  nm, forgot to update PlatformToolset :)
>
> I find myself doing that everytime I do a build. Maybe this PR is a good time to udate the default?

sipsorcery commented:
reACK 05b2d9fe076233961ebf24e74d47c57cc0618bbb.

jarolrod reviewed (commented):
ACK 05b2d9fe076233961ebf24e74d47c57cc0618bbb

Changes look good to me, makes sense to update as visual studio 2019 will "will go out of support in October 2022". PR description could state some of the motivation for the update.

hebasto commented:
> ACK [05b2d9f](https://github.com/bitcoin/bitcoin/commit/05b2d9fe076233961ebf24e74d47c57cc0618bbb)
> 
> Changes look good to me, makes sense to update as visual studio 2019 will "will go out of support in October 2022". PR description could state some of the motivation for the update.

PR description has been just added.
